utils/upload.go

```diff
@@ -0,0 +1,180 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+
+	"cloud.google.com/go/storage"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+// channel to extract files from the folder
+type fileWalk chan string
+
+type Uploader interface {
+	Upload(walker fileWalk)
+}
+
+func UploadToCloudStorage(uploader Uploader, path string) {
+	walker := make(fileWalk)
+	go func() {
+		//get files to upload via the channel
+		if err := filepath.Walk(path, walker.WalkFunc); err != nil {
+			log.Fatalln("Walk failed: ", err)
+		}
+
+		close(walker)
+
+	}()
+
+	uploader.Upload(walker)
+
+}
+
+func (f fileWalk) WalkFunc(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if !info.IsDir() {
+		f <- path
+	}
+
+	return nil
+}
+
+type AwsUploader struct {
+	BucketName      string
+	Prefix          string
+	Region          string
+	FolderLocalPath string
+}
+
+func (a AwsUploader) Upload(walker fileWalk) {
+
+	//creating a new session
+	sess, err := session.NewSession(&aws.Config{
+		Region:                        aws.String(a.Region),
+		CredentialsChainVerboseErrors: aws.Bool(true),
+	})
+	if err != nil {
+		log.Fatalf("failed to create a session")
+	}
+
+	bucket := a.BucketName
+	log.Printf("bucket %s", bucket)
+
+	prefix := a.Prefix
+
+	uploader := s3manager.NewUploader(sess)
+	for path := range walker {
+		// rel, err := filepath.Rel(a.FolderLocalPath, path)
+		// if err != nil {
+		// 	log.Fatalln("Unable to get relative path ", a.FolderLocalPath, path)
+		// }
+		filename := filepath.Base(path)
+
+		file, err := os.Open(path)
+		if err != nil {
+			log.Println("Failed opening file", path, err)
+			continue
+		}
+
+		defer file.Close()
+
+		result, err := uploader.Upload(&s3manager.UploadInput{
+			Bucket: &bucket,
+			Key:    aws.String(filepath.Join(prefix, filename)),
+			Body:   file,
+		})
+
+		if err != nil {
+			log.Fatalln("Failed to upload", path, err)
+		}
+		log.Println("Uploaded", path, result.Location)
+	}
+}
+
+type GcpUploader struct {
+	BucketName string
+	ProjectId  string
+	UploadPath string
+}
+
+func (g *GcpUploader) Upload(walker fileWalk) {
+	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", os.Getenv("GCP_CREDENTIALS"))
+	client, err := storage.NewClient(context.Background())
+	if err != nil {
+		log.Fatalln("Failed to create client ", err)
+	}
+	for path := range walker {
+		filename := filepath.Base(path)
+		fmt.Printf("Creating file /%v/%v\n", g.BucketName, filename)
+
+		ctx := context.Background()
+
+		wc := client.Bucket(g.BucketName).Object(g.UploadPath + filename).NewWriter(ctx)
```

Comment:
Same bucket name here as well, `constants.S3_BUCKET_NAME`

Reply:
this is the bucket name for the google cloud provider. why do we use the S3_BUCKET_NAME here?

Reply:
Yeah you can create a new constant for GCP and Azure and use it